### PR TITLE
fix(Bun.Socket) onDrain will only be called after handshake + docs update

### PR DESCRIPTION
**Pull request by cirospaciari**
…ome docs update

### What does this PR do?

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

added a test

### CODE REVIEW

**Comment by Jarred-Sumner** [comment]
This should be at the uSockets layer right? If SSL handshaking is in progress the socket isn’t truly writable

**Comment by cirospaciari** [comment]
> This should be at the uSockets layer right? If SSL handshaking is in progress the socket isn’t truly writable

you are right, done in usockets layer now

File: src/bun.js/api/bun/socket.zig [hunk 1]
```diff
@@ -1519,6 +1519,9 @@ fn NewSocket(comptime ssl: bool) type {
             // you should use getAuthorizationError and authorized getter to get those values in this case
             if (is_open) {
                 result = callback.callWithThis(globalObject, this_value, &[_]JSValue{this_value});
+                // clean onOpen callback so only called in the first handshake and not in every renegotiation
+                this.handlers.onOpen.unprotect();
```
Comment (Jarred-Sumner) [inline src/bun.js/api/bun/socket.zig]: This will break for servers

Comment (cirospaciari) [reply src/bun.js/api/bun/socket.zig]: fixed

File: test/js/bun/net/socket.test.ts [hunk 2]
```diff
@@ -477,3 +477,24 @@ it("should connect directly when using an ip address", async () => {
...
+          reject(new Error("Socket not authorized"));
+        }
+      },
+      handshake() {
+        resolve();
+      },
+    },
+  });
+  await promise;
+  expect(socket.authorized).toBe(true);
```
Comment (Jarred-Sumner) [inline test/js/bun/net/socket.test.ts:499]:
```suggestion
  expect(socket.authorized).toBe(true);
```

File: test/js/bun/net/socket.test.ts [hunk 3]
```diff
@@ -477,3 +477,24 @@ it("should connect directly when using an ip address", async () => {
   await promise;
   expect(opened).toBe(true);
 });
+
+it("should not call drain before handshake", async () => {
+  const { promise, resolve, reject } = Promise.withResolvers();
+  const socket = await Bun.connect({
```
Comment (Jarred-Sumner) [inline test/js/bun/net/socket.test.ts]:
```suggestion
  using socket = await Bun.connect({
```

**Pull request merged**